### PR TITLE
Fix BadProperty checks

### PR DESCRIPTION
**Pull request by akorgor**
This PR introduces additional `BadProperty` assertions based on restrictions on the parameters from the literature. Furthermore, it makes the format of the thrown messages consistent across all the files. Priority there was to make it as easy for the user as possible to fix a badly set parameter. For this reasons the parameter name (e.g.`tau_m`) instead of the description (e.g. "membrane time constant") is written in the message as well as a positive formulation ("must be ..." instead of "must not be ...").

### CODE REVIEW

File: models/eprop_iaf_psc_delta_adapt.cpp [hunk 1]
```diff
@@ -158,47 +160,65 @@ nest::eprop_iaf_psc_delta_adapt::Parameters_::set( const DictionaryDatum& d, Nod
...
+  }
   if ( f_target_ < 0 )
   {
-    throw BadProperty( "Target firing rate must be >= 0." );
+    throw BadProperty( "f_target must be >= 0." );
+  }
+
+  if ( gamma_ < 0.0 or 1.0 <= gamma_ )
+  {
+    throw BadProperty( "gamma must be in [0,1)." );
```
Comment (JesusEV) [inline models/eprop_iaf_psc_delta_adapt.cpp:189]: @akorgor  Where does this condition come from?

File: models/eprop_synapse.h [hunk 2]
```diff
@@ -476,14 +476,24 @@ eprop_synapse< targetidentifierT >::set_status( const DictionaryDatum& d, Connec
   updateValue< double >( d, names::adam_m, adam_m_ );
   updateValue< double >( d, names::adam_v, adam_v_ );
 
-  if ( weight_ < Wmin_ or weight_ > Wmax_ )
+  if ( eta_ < 0 )
   {
-    throw BadProperty( "Wmin <= weight <= Wmax must be satisfied." );
+    throw BadProperty( "eta must be >= 0." );
   }
```
Comment (JesusEV) [inline models/eprop_synapse.h]: @akorgor Where does this condition come from?

File: models/eprop_iaf_psc_delta_adapt.cpp (hunk 1 above)
Comment (akorgor) [reply models/eprop_iaf_psc_delta_adapt.cpp:189]: p. 2 in Supplemental of "Long short-term memory and Learning-to-learn in networks of spiking neurons" says it should be < 1. Since it is referred to as an "amplitude", it implies that it's not negative.

File: models/eprop_synapse.h (hunk 2 above)
Comment (akorgor) [reply models/eprop_synapse.h]: Learning rates are generally considered to be positive. As far as I understand, we would also move away from the optimum with a negative learning rate.

File: models/eprop_iaf_psc_delta_adapt.cpp (hunk 1 above)
Comment (JesusEV) [reply models/eprop_iaf_psc_delta_adapt.cpp:189]: Thanks for locating this. What catches my attention is the way $v_j(t)$ is defined immediately after eq.6, which is slightly different from what we do in the code (the denominator specifically ).

File: models/eprop_synapse.h (hunk 2 above)
Comment (JesusEV) [reply models/eprop_synapse.h]: In pytorch there is the "maximize" flag when one is interested in maximizing the loss function instead of minimizing it. Since we do not have such a flag, it seems convenient for the user to simply set the learning rate negative in case they want to do gradient ascent. Another example where it is useful is in adversarial networks, where two networks, or two different sections of the same network, have opposite objectives and one tries to maximize the loss while the other tries to minimize it.

**Comment by JesusEV** [comment]
I find the proposed changes very useful and well formulated. Apart from giving the user the freedom to choose the learning rate as they wish, it seems to me that the overall changes should be incorporated as soon as possible.

File: models/eprop_synapse.h (hunk 2 above)
Comment (akorgor) [reply models/eprop_synapse.h]: Nice points - convinced me and I will remove this restriction.

File: models/eprop_iaf_psc_delta_adapt.cpp (hunk 1 above)
Comment (akorgor) [reply models/eprop_iaf_psc_delta_adapt.cpp:189]: Oh, nice catch. Any effect of the different denominator inside the bracket on the gamma is probably swallowed by the $abs()$ and $\text{max}()$. And, yes, the denominator $v_\text{th}$ in the beginning, which is missing in the reference above, could lead to an overall negative prefactor that possibly goes beyond 1 (since $v_\text{th}$ is sometimes negative and > 1). Concretely, if a user would like to reproduce the pseudo-derivative in the reference by choosing $\gamma=0.3\cdot v_\text{th}$, the assert would prevent it. Considering this, I am happy to lift both restrictions on this hyperparameter if you agree.

Comment (akorgor) [reply models/eprop_iaf_psc_delta_adapt.cpp:189]: However, since it got so easy to implement a new pseudo derivative in our code, I think, it would be cleanest to implement in case of a use-case for the pseudo derivative in the reference a new pseudo derivative function (which will simplify even further as soon as NESTML will be able to generate it). Hence, I would keep the restriction on gamma here.

Comment (JesusEV) [reply models/eprop_iaf_psc_delta_adapt.cpp:189]: The topic of absorptions goes further. Since the filter $\mathcal F_\kappa()$ does not create powers of $\psi$, only powers of $\kappa$ multiplying the sequence of $\psi$'s, the factor $\gamma$, (or, if you want $\gamma/ v_{\text{th}}$) can be factored out, and absorbed into the learning rate. In fact, if you set `gamma=1.0` and `eta=1e-4 * 0.3` for the `in-rec` and `rec-rec` synapses, the verification check is successful.

The fact that these factors can be absorbed by the learning rate (exactly for GD and approximately for Adam), makes me think that the constraint $\gamma < 1$ is not as restrictive as it might appear. Users can modify the learning rate to achieve effects similar to increasing $\gamma$ beyond 1. 

Therefore it seems to me that the restriction could be maintained to keep consistency with the reference you mention (even when specified for a different definition of $\psi$).
